vllm/core/block_manager_v1.py

```diff
@@ -559,15 +651,32 @@ def free(self, seq: Sequence) -> None:
         self._free_block_table(block_table)
         del self.block_tables[seq.seq_id]
 
+    def free_cross(self, seq_group: SequenceGroup) -> None:
```

Comment:
I what situation would there be a need to free the cross attention blocks independently of the self attention blocks?

Reply:
Hi @maxdebayser , Technically, there are times when the scheduler frees individual decoder sequences, however you are right that encoder sequences are only freed as part of a group, alongside decoder sequences.

The need for free_cross(SequenceGroup) primarily results from my choice to identify encoder sequences by their SequenceGroup's request id, which I chose to do because each SequenceGroup has <= 1 encoder sequence.

Reply:
Thanks for the explanation. It absolutely makes sense to me to identify the encoder sequence by the request ID. So would the `free_cross` method be called from `Scheduler.free_finished_seq_groups()`?

Reply:
Hmm. It looks like Scheduler.free_finished_seq_groups() is dequeueing SequenceGroups into the self.running member?

I think the most likely place would be

Scheduler.abort_seq_group()

Reply:
Right, It would have to be called there too.

Sorry, I know this is part of the next PR :D, but in LLMEngine._process_model_outputs() the output processors are called on each seq in the seq group and scheduler.free_seq() is called on the finished ones. After that _process_model_outputs() calls free_finished_seq_groups(), so this could be a convenient place to call free_cross() on all finished seq groups before the references are lost.

Reply:
Oh I see, noted. Thanks!